<a href="https://colab.research.google.com/github/hush-cz/ML_exp_tr_v2/blob/main/Fine_tuning_DistilBERT_for_Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sentiment Analysis with BERT on IMDB Dataset

**Author:** Tomas Vodacek
**Goal:** Fine-tune a pre-trained DistilBERT model to classify IMDB movie reviews as positive or negative.  
**Dataset:** IMDB (25k labeled reviews).  
**Model:** DistilBERT from HuggingFace.  

In [ ]:
!pip uninstall -y transformers
!pip install --no-cache-dir transformers==4.57.0 datasets accelerate scikit-learn
!pip install pyarrow==21.0.0

Found existing installation: transformers 4.56.2
Uninstalling transformers-4.56.2:
  Successfully uninstalled transformers-4.56.2
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.4/41.4 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 191.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 MB 14.3 MB/s eta 0:00:00
  Attempting uninstall: pyarrow
    Found existing installation: pyarrow 18.1.0
    Uninstalling pyarrow-18.1.0:
      Successfully uninstalled pyarrow-18.1.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.
pylibcudf-cu12 25.6.0 requires pyarrow<20.0.0a0,>=14.0.0; platform_machine == "x86_64", but you have pyarrow 21.0.0 which is incompatible.


In [ ]:
from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer
import numpy as np
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
from google.colab import userdata
userdata.get('HF_TOKEN')

from huggingface_hub import login
import os

if "HF_TOKEN" in os.environ:
    login(os.environ["HF_TOKEN"])
else:
    print("⚠️ HF_TOKEN not found. Add it in Colab Secrets (Tools → Secrets).")

⚠️ HF_TOKEN not found. Add it in Colab Secrets (Tools → Secrets).


In [ ]:
# Načti IMDB dataset
dataset = load_dataset("imdb")

# Tokenizace textu
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

def tokenize(batch):
    return tokenizer(batch["text"], padding=True, truncation=True)

dataset = dataset.map(tokenize, batched=True)
dataset = dataset.rename_column("label", "labels")
dataset.set_format("torch", columns=["input_ids", "attention_mask", "labels"])

README.md: 0.00B [00:00, ?B/s]

plain_text/train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

plain_text/test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

plain_text/unsupervised-00000-of-00001.p(…):   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [ ]:
model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=2)

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="./results",
    eval_strategy="steps",   # vyhodnocování po krocích
    eval_steps=500,                # jak často vyhodnocovat
    save_strategy="steps",         # ukládání také po krocích
    save_steps=500,
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=1,
    weight_decay=0.01,
    logging_dir="./logs"
)

In [ ]:
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    accuracy = accuracy_score(labels, predictions)
    return {"accuracy": accuracy}

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["test"],
    compute_metrics=compute_metrics,
)

In [ ]:
# ✅ Načtení knihoven
from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer
from sklearn.metrics import accuracy_score
import numpy as np
import torch
import os
from transformers.trainer_utils import IntervalStrategy # Import IntervalStrategy

# ✅ Načtení datasetu IMDB
dataset = load_dataset("imdb")

# ✅ Inicializace tokenizeru a modelu
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")
model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=2)

# ✅ Tokenizace (převod textů na čísla)
def tokenize_function(example):
    return tokenizer(example["text"], padding="max_length", truncation=True)

tokenized_datasets = dataset.map(tokenize_function, batched=True)

# ✅ Vytvoření menšího datasetu (rychlý trénink)
small_train = tokenized_datasets["train"].shuffle(seed=42).select(range(8000))
small_test = tokenized_datasets["test"].shuffle(seed=42).select(range(2000))

# ✅ Funkce pro výpočet metrik
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    accuracy = accuracy_score(labels, predictions)
    return {"accuracy": accuracy}

# ✅ Nastavení tréninku
os.environ["WANDB_DISABLED"] = "true"  # vypne W&B přihlášku

training_args = TrainingArguments(
    output_dir="./results",
    eval_strategy=IntervalStrategy.STEPS, # Use IntervalStrategy.STEPS
    eval_steps=500,
    save_strategy=IntervalStrategy.STEPS, # Use IntervalStrategy.STEPS
    save_steps=500,
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    num_train_epochs=6,
    weight_decay=0.01,
    load_best_model_at_end=True,
    logging_dir="./logs",
    logging_steps=100,
)

# ✅ Vytvoření trenéra
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=small_train,
    eval_dataset=small_test,
    compute_metrics=compute_metrics,
)

# ✅ Spuštění tréninku
trainer.train()

# ✅ Vyhodnocení výsledků
metrics = trainer.evaluate()
print("Evaluation metrics:", metrics)

# ✅ Krátký test modelu na vlastních větách
texts = [
    "This movie was amazing, I loved every minute!",
    "It was the worst film I have ever seen."
]

inputs = tokenizer(texts, padding=True, truncation=True, return_tensors="pt")
inputs = {k: v.to(model.device) for k, v in inputs.items()}
outputs = model(**inputs)
preds = np.argmax(outputs.logits.detach().cpu().numpy(), axis=1)

for text, pred in zip(texts, preds):
    sentiment = "Positive" if pred == 1 else "Negative"
    print(f"{sentiment}: {text}")

README.md: 0.00B [00:00, ?B/s]

plain_text/train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

plain_text/test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

plain_text/unsupervised-00000-of-00001.p(…):   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/50000 [00:00<?, ? examples/s]

Using the `WANDB_DISABLED` environment variable is deprecated and will be removed in v5. Use the --report_to flag to control the integrations used for logging result (for instance --report_to none).


Evaluation metrics: {'eval_loss': 0.2381664514541626, 'eval_accuracy': 0.911, 'eval_runtime': 30.6204, 'eval_samples_per_second': 65.316, 'eval_steps_per_second': 8.164, 'epoch': 6.0}
Positive: This movie was amazing, I loved every minute!
Negative: It was the worst film I have ever seen.


In [ ]:
import transformers
print(transformers.__file__)
help(transformers.TrainingArguments)

/usr/local/lib/python3.12/dist-packages/transformers/__init__.py
Help on class TrainingArguments in module transformers.training_args:

class TrainingArguments(builtins.object)
 |  TrainingArguments(output_dir: Optional[str] = None, overwrite_output_dir: bool = False, do_train: bool = False, do_eval: bool = False, do_predict: bool = False, eval_strategy: Union[transformers.trainer_utils.IntervalStrategy, str] = 'no', prediction_loss_only: bool = False, per_device_train_batch_size: int = 8, per_device_eval_batch_size: int = 8, per_gpu_train_batch_size: Optional[int] = None, per_gpu_eval_batch_size: Optional[int] = None, gradient_accumulation_steps: int = 1, eval_accumulation_steps: Optional[int] = None, eval_delay: Optional[float] = 0, torch_empty_cache_steps: Optional[int] = None, learning_rate: float = 5e-05, weight_decay: float = 0.0, adam_beta1: float = 0.9, adam_beta2: float = 0.999, adam_epsilon: float = 1e-08, max_grad_norm: float = 1.0, num_train_epochs: float = 3.0, max_steps: 

In [ ]:
metrics = trainer.evaluate()
print("Evaluation metrics:", metrics)

Evaluation metrics: {'eval_loss': 0.2381664514541626, 'eval_accuracy': 0.911, 'eval_runtime': 33.2949, 'eval_samples_per_second': 60.069, 'eval_steps_per_second': 7.509, 'epoch': 6.0}


In [ ]:
print(metrics)

{'eval_loss': 0.2381664514541626, 'eval_accuracy': 0.911, 'eval_runtime': 33.2949, 'eval_samples_per_second': 60.069, 'eval_steps_per_second': 7.509, 'epoch': 6.0}


In [ ]:
sample_texts = [
    "This movie was fantastic! I loved it.",
    "The film was terrible and boring."
]

inputs = tokenizer(sample_texts, padding=True, truncation=True, return_tensors="pt")
# Move inputs to the same device as the model
inputs = {k: v.to(model.device) for k, v in inputs.items()}
outputs = model(**inputs)
preds = np.argmax(outputs.logits.detach().cpu().numpy(), axis=1) # Move logits back to CPU for numpy

for text, label in zip(sample_texts, preds):
    sentiment = "Positive" if label == 1 else "Negative"
    print(f"Review: {text}\nPrediction: {sentiment}\n")

Review: This movie was fantastic! I loved it.
Prediction: Positive

Review: The film was terrible and boring.
Prediction: Negative





## Results and Discussion

During the fine-tuning of the **DistilBERT** model on the **IMDB movie reviews dataset**, a steady improvement in classification accuracy was observed across the training epochs.  
After six epochs, the model achieved an **evaluation loss of 0.238** and an **accuracy of 0.911**, corresponding to approximately **91% correct sentiment classification**.  
Training was performed using GPU acceleration in Google Colab, with an average evaluation speed of about **65 samples per second**.

The model correctly classified the following test sentences:  
- *"This movie was amazing, I loved every minute!"* → **Positive**  
- *"It was the worst film I have ever seen."* → **Negative**  
demonstrating its ability to capture emotional tone and textual polarity even with a relatively small training subset.

The results show that the lightweight **DistilBERT** architecture provides an excellent balance between performance and computational efficiency.  
Compared to the full BERT model, DistilBERT significantly reduces training time and memory usage while maintaining high accuracy (with only a 2–3% drop on average).  
The model’s performance could be further improved by expanding the dataset, increasing the number of epochs, or fine-tuning hyperparameters such as the **learning rate** and **batch size**.

Overall, this experiment confirms that **fine-tuning pretrained Transformer models** is a highly effective approach for text classification tasks, even with limited data.  
The project demonstrates a practical application of modern **Natural Language Processing (NLP)** methods and can serve as a foundation for further experiments in **sentiment analysis**, **emotion detection**, or **automated review evaluation**.



## Conclusion

In conclusion, this project successfully demonstrated the process of fine-tuning a pretrained Transformer model (**DistilBERT**) for sentiment analysis.  
With just a limited subset of the IMDB dataset and a modest number of training epochs, the model achieved strong performance, reaching over **91% accuracy** in classifying movie reviews as positive or negative.  
The experiment highlights the power and efficiency of modern NLP architectures, showing that meaningful text understanding can be achieved even with lightweight models and accessible computational resources.  
This fine-tuned model represents a solid foundation for further exploration in applied natural language processing — including **emotion recognition**, **customer feedback analysis**, or **automated review systems**.

